RFCs/rfc-0010-wallet-provider-attestation-api.md

```diff
@@ -0,0 +1,252 @@
+    REQUEST[1] attestation
+    IF user approves
+        RESOLVE[2] attestation
+        CONTINUE zkProvableProgram
+    IF user rejects
+        REJECT[3] attestation
+        STOP interaction
+IF provider is undefined
+    STOP interaction
+```
+
+[1] REQUEST
+The requester MUST request attestation by calling the `mina_requestAttestation` RPC method on the provider exposed at `window.mina`. Calling this method MUST trigger a user interface allowing the user to approve or reject the attestation request. This method MUST return a Promise that is resolved with the desired proof or rejected if the attestation cannot be produced.
+
+[2] RESOLVE
+The Promise returned when calling the `mina_requestAttestation` RPC method MUST be resolved with the desired proof. This should emit an event message indicating that the attestation request has been approved and the promise has been resolved with the desired proof.
+
+[3] REJECT
+The Promise returned when calling the `mina_requestAttestation` RPC method MUST be rejected with an informative `Error` if the attestation cannot be produced. This should emit an event message indicating that the attestation request has been denied, and the promise is rejected with an error.
+
+When a requester seeks attestation from a wallet, they must provide contextual bindings. These bindings are key-value pairs where the keys are strings, and the values can be any type with valid JSON encoding.
+
+#### `mina_requestAttestation` Binding Parameters
+
+#### `program`
+A unique identifier or hash representing the provable program the wallet should run internally for attestation. This might correspond to the verification key of the provable program or its hash.
+
+### `clientData`
+An adaptable data structure designed to bind the wallet to contextually relevant information. The content and structure of `clientData` are dynamic, tailored to the specific scenario or attestation requirement at hand. It can encapsulate a range of data, from verifiable credentials to custom datasets, based on what the wallet is expected to know when attesting. Given the myriad potential use cases, the shape and constituents of clientData can vary widely. Requesters should structure it based on prior knowledge of the data the wallet is acquainted with, ensuring effective communication and validation during the attestation process.
+
+#### `publicInputs`
+A key-value map detailing the provable-program's method signature for open inputs. Each key-value pair describes a specific input and the value to use in the provable-program; further binding the wallet.
+
+
+#### Example On-Chain Usage -- request attestation using a provable program for royalties claim
+```ts
+import { Mina } from "o1js";
+
+const mina = window.mina;
+
+try {
+    // Define the binding for the attestation
+    const binding = {
+        // The program could be constrained by a unique identifier for the program like a verification key
+        program: "ProvableProgramForRoyaltyClaim",
```

Comment:
How does this program selection work? Is it something already hardcoded into the wallet, or does it make it in another way?

If hardcoded, maybe we should consider something that operates over json records? Like instead of a string being able to specify proofs over statements on records? (Like I think we talked about at an early version of this, getting a field and proving its type and basic things about it like for strings checking if it matches another string, for numbers inequalities, etc)

Reply:
The `binding` is part of the request the application makes to the wallet. So the `program` field tells the wallet which provable program to use, this naively assumes the wallet knows what program that is and it's background script can fetch/compile the program such that it can execute the program with the defined inputs (public or private).

This certainly needs inputs from wallet providers to understand what is possible to implement. Holistically, I see these provable programs as little zkApps that are modules in the wallet that can be added & removed.

Reply:
I would propose then,we do something that operates over records, so that people can add new attestations without needing to re-coordinate and get updates from wallet providers? I think we should add that to this spec as well (lmk, if needed I can find the older doc I think we had on it too somewhere...)

Reply:
Thinking about this a bit more, I think that one option could be:

- Have attestations be json records (with extra types for different crypto primitives; and maybe flat json records)
- Make a DSL that operates over json records, and can be safely sent to the wallet to be executed in the wallet's context without security risks
- Have the DSL able to provide the following:
    - Binary statement over whether an attestation passes the check performed by the DSL, to compute the proof
        - this can also check which attestations would pass the DSL, to ask the user which to use if they have multiple attestations that would pass (or let them know if they have none)
    - Detail so the wallet can notify the user what information the DSL is asking for from the attestation (eg this will check [ Age > 20, proof of unique identity wrt nullifier X, etc ]

The DSL could be like is done for o1js, where its just a typescript library;

Maybe something like 
```
(attestation) => {
   const verifier = new AttestationVerifier(attestation);
   return verifier.check_equal("field1", "value")
              .and(verifier.check_signature("field2", verifier.get("data_field"), "publicKey"))
              .and(verifier.check_gt("field3", 20));
}
```

I think ideally we'd want an ecosystem of attestations, so hopefully this could help make it easier to produce / consume attestations permissionlessly. Curious on any thoughts!

Reply:
The `bindings` should satisfy those requirements -- they're about constraining the wallet to attest to something specific:
- Which provable-program to use
- Who the expected issuer of the private data is
- What the expected logic gate is (e.g. `age > 20`)

Since a wallet at it's core is just a key management system, it is not impossible to imagine an attestation management system that can sit next to the KMS or integrate into it. Much like a generic `api.sign(params)` can be one API of that key management system `api.prove(params)` could be too. That `params` argument is the `bindings` and should serve the purpose as might a `api.sign(TransactionBody)` might. Wallet users can already attest to statements in the browser context like constructing transactions (they are the spender/fee payer of that transaction), only to pass to the wallet for signing/submission, the key difference with this attestation API is the wallet is asked to use the `bindings` as context for producing the required proof within its own application.

In that sense, the attestation API _could_ be used as a way for a zkApp to give context to the wallet about how to construct a transaction, sign it, and prove it before submitting to the network. I drew a little diagram to show the shift in responsibilities:
<img width="607" alt="Screenshot 2024-01-18 at 10 32 25" src="https://github.com/MinaFoundation/Core-Grants/assets/92999717/4ee2e344-982d-47f9-a3ed-a51ff83fe35e">

For this attestation context it is about giving the wallet context on how to construct arguments for a provable-program method and prove it.

Reply:
What does bindings look like exactly in this model? I'm still a little confused exactly where the proof if computed, and how exactly it is computed?

I was thinking it probably needs to be computed in the wallet, so that the attestation data doesn't leave the wallet to the unsafe webpage, but then the wallet needs some specific way of knowing how to compute the proof.

Reply:
In particular, we say above the wallet would be told "Which provable-program to use" - to me that sounds like wallets would have a fixed set of programs they can use - which seems limiting if we want new attestation producers and consumers to come along and build new attestations and consume them in their own way.

Reply:
> What does bindings look like exactly in this model?

The bindings are [defined here](https://github.com/MinaFoundation/Core-Grants/blob/rfc-0010-wallet-provider-private-attestation-api/RFCs/rfc-0010-wallet-provider-attestation-api.md#mina_requestattestation-binding-parameters). 

> where the proof if computed, and how exactly it is computed?

The bindings define what program and what inputs are expected via `ClientData` and `publicInputs`, proving should happen in the wallet application (or any secure enclave). `bindings` is the data structure that this RFC needs the most feedback on as it defines what `provable-program` and what inputs the wallet should use. For example, a wallet should be bound (hence the name `bindings`) to prove a specific `Issuer` created a credential for their `PublicKey`; they shouldn't be able to prove that _any_ `Issuer` created that credential.

> "Which provable-program to use" to me that sounds like wallets would have a fixed set of programs they can use 

It should be read as "which provable-program to use for this attestation". There is no limit on the set of provable-programs a wallet can know about. As long as a wallet can know (a.) where to obtain the specific program, (b.) compile it, and (c.) cache the program to prevent always re-compiling when attesting, there isn't a limitation (modulo browser/service worker constraints).

> I think we would still need something like the above DSL, so that we can allow dynamic proofs over new

The attestations created can be stored as credentials in the wallet's storage for re-use, but this somewhat can conflict with the considerations section where attestations may require a TTL.

Reply:
Ah I see - so can the bindings can define a series of operations constituting a claim (which could be constructed via DSL, or could be statically defined as done here)?

I think then I'm not understanding the role of the provable-program? Like what does changing that done?

I think I'm proposing then, in this new context, to
- Only provide information on the claim, as a statement over the fields of attestations (and not have provable-program? Unless I'm probably missing why its necessary)
- Maybe use a DSL to construct claims to make it easier? Though probably not necessary, especially for a first version.

Reply:
Just to share context here from speaking offline, we met and confirmed it does make sense to have "provable-programs" - but that provable-programs should be probably added via the standards process, and this standard should include a first provable program, “RecordCredentialAttestation” or something of the like, that operates with boolean statements over structured json-like attestation objects. If I have time later, I’ll make a PR on this one with what that could look like.